scanpy/get/_aggregated.py

```diff
@@ -0,0 +1,437 @@
+        nonetheless be non-zero. By default, our implementation resets the variance to exactly zero
+        when the computed variance, relative to the squared mean, nears limit of precision of the
+        floating-point significand.
+
+        Params
+        ------
+        dof
+            Degrees of freedom for variance.
+
+        Returns
+        -------
+        dict with mean, count, and var keys.
+        """
+        assert dof >= 0
+        A, _ = self._sparse_aggregator(normalize=True)
+        count_ = np.bincount(self._key_index)
+        mean_ = utils.asarray(A @ self._data)
+        mean_sq = utils.asarray(A @ _power(self._data, 2))
+        if self._weight is None:
+            sq_mean = mean_**2
+        else:
+            A_unweighted, _ = Aggregate(
+                groupby=self._groupby,
+                data=self._data,
+                weight=self._weight,
+                key_set=self._key_set,
+            )._sparse_aggregator()
+            mean_unweighted = utils.asarray(A_unweighted * self._data)
+            sq_mean = 2 * mean_ * mean_unweighted + mean_unweighted**2
+        var_ = mean_sq - sq_mean
+        precision = 2 << (42 if self._data.dtype == np.float64 else 20)
+        # detects loss of precision in mean_sq - sq_mean, which suggests variance is 0
+        var_[precision * var_ < sq_mean] = 0
+        if dof != 0:
+            var_ *= (count_ / (count_ - dof))[:, np.newaxis]
+        return {'mean': mean_, 'var': var_, 'count': count_}
+
+    def _sparse_aggregator(
+        self, normalize: bool = False
+    ) -> Tuple[coo_matrix, np.ndarray]:
+        """
+        Form a coordinate-sparse matrix A such that rows of A * X
+        are weighted sums of groups of rows of X.
+
+        A[i, j] = w includes X[j,:] in group i with weight w.
+
+        Params
+        ------
+        normalize
+            If true, weights for each group are normalized to sum to 1.0,
+            corresponding to (weighted) mean.
+
+        Returns
+        -------
+        A
+            weighted sums of groups of rows of X.
+        keys
+            An ndarray with keys[i] the group key corresponding to row i of A.
+        """
+        keys, key_index, df_index, weight_value = self._extract_indices()
+        if df_index is None:
+            df_index = np.arange(len(key_index))
+        if self._weight is None:
+            weight_value = np.ones(len(key_index))
+        A = coo_matrix(
+            (weight_value, (key_index, df_index)),
+            shape=(len(keys), self._data.shape[0]),
+        )
+        if normalize:
+            n_row = A.shape[0]
+            row_sums = np.asarray(A.sum(axis=1))
+            D = dia_matrix(((row_sums.T**-1), [0]), shape=(n_row, n_row))
+            A = D * A
+        return A, keys
+
+    def _extract_indices(self):
+        def _filter_indices(key_set, keys, key_index, df_index, weight_value=None):
```

Comment:
please explain on a high level what these do (in comments or docstring)

Reply:
Ditto for old code.  Will review this and add docstrings.